File: server/app/routers/widgets.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from server.app.database import get_db
from server.app.models import Layout, Widget
from server.app.schemas import WidgetCreate, WidgetUpdate, WidgetResponse, WidgetPositionUpdate

router = APIRouter(tags=["widgets"])


@router.post("/api/layouts/{layout_id}/widgets", status_code=201, response_model=WidgetResponse)
def add_widget(layout_id: int, body: WidgetCreate, db: Session = Depends(get_db)):
    layout = db.query(Layout).filter(Layout.id == layout_id).first()
    if not layout:
        raise HTTPException(status_code=404, detail="Layout not found")
    widget = Widget(
        layout_id=layout_id,
        widget_type=body.widget_type,
        config=body.config,
        position_x=body.position_x,
        position_y=body.position_y,
        width=body.width,
        height=body.height,
    )
    db.add(widget)
    db.commit()
    db.refresh(widget)
    return widget


@router.put("/api/widgets/{widget_id}", response_model=WidgetResponse)
def update_widget(widget_id: int, body: WidgetUpdate, db: Session = Depends(get_db)):
    widget = db.query(Widget).filter(Widget.id == widget_id).first()
    if not widget:
        raise HTTPException(status_code=404, detail="Widget not found")
    update_data = body.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(widget, key, value)
    db.commit()
    db.refresh(widget)
    return widget


@router.delete("/api/widgets/{widget_id}", status_code=204)
def delete_widget(widget_id: int, db: Session = Depends(get_db)):
    widget = db.query(Widget).filter(Widget.id == widget_id).first()
    if not widget:
        raise HTTPException(status_code=404, detail="Widget not found")
    db.delete(widget)
    db.commit()


@router.put("/api/layouts/{layout_id}/widgets/positions", response_model=list[WidgetResponse])
def batch_update_positions(
    layout_id: int, positions: list[WidgetPositionUpdate], db: Session = Depends(get_db)
):
    layout = db.query(Layout).filter(Layout.id == layout_id).first()
    if not layout:
        raise HTTPException(status_code=404, detail="Layout not found")
    widgets = []
    for pos in positions:
        widget = db.query(Widget).filter(Widget.id == pos.id, Widget.layout_id == layout_id).first()
        if not widget:
            raise HTTPException(status_code=404, detail=f"Widget {pos.id} not found in layout")
        widget.position_x = pos.position_x
        widget.position_y = pos.position_y
        widget.width = pos.width
        widget.height = pos.height
        widgets.append(widget)
    db.commit()
    for w in widgets:
        db.refresh(w)
    return widgets

File: src/server/routes/widgets.ts
import type { FastifyInstance } from 'fastify'
import type Database from 'better-sqlite3'
import { requireAuth } from '../middleware/auth.js'
import {
  createWidget,
  getWidget,
  updateWidget,
  deleteWidget,
  batchUpdatePositions,
  getWidgetsByLayout,
} from '../db/queries/widgets.js'
import { getLayout } from '../db/queries/layouts.js'
import { geocodeZip, GeocodingError } from '../services/geocoding.js'
import {
  WidgetCreateSchema,
  WidgetUpdateSchema,
  WidgetPositionUpdateSchema,
} from '@shared/types.js'
import type { WidgetCreate, WidgetUpdate, WidgetPositionUpdate } from '@shared/types.js'
import { Type } from '@sinclair/typebox'

async function resolveWeatherZip(
  config: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  if (!('zip_code' in config)) {
    return config
  }
  try {
    const geo = await geocodeZip(config.zip_code as string)
    return { ...config, lat: geo.lat, lon: geo.lon, location_name: geo.locationName }
  } catch (e) {
    if (e instanceof GeocodingError) {
      throw e
    }
    throw e
  }
}

export async function widgetRoutes(app: FastifyInstance): Promise<void> {
  const db = (app as unknown as { db: Database.Database }).db

  app.post<{ Params: { layoutId: string }; Body: WidgetCreate }>(
    '/api/layouts/:layoutId/widgets',
    {
      schema: { body: WidgetCreateSchema },
      preHandler: [requireAuth],
    },
    async (request, reply) => {
      const layoutId = Number(request.params.layoutId)
      const layout = getLayout(db, layoutId)
      if (!layout) {
        reply.code(404).send({ error: 'Layout not found' })
        return
      }

      let config = request.body.config
      if (request.body.widget_type === 'weather') {
        try {
          config = await resolveWeatherZip(config)
        } catch (e) {
          if (e instanceof GeocodingError) {
            reply.code(400).send({ error: e.message })
            return
          }
          throw e
        }
      }

      const widget = createWidget(db, layoutId, { ...request.body, config })
      app.log.info(
        `Widget added: ${widget.widget_type} (id=${widget.id}) to layout ${layoutId}`,
      )
      reply.code(201).send(widget)
    },
  )

  app.put<{ Params: { id: string }; Body: WidgetUpdate }>(
    '/api/widgets/:id',
    {
      schema: { body: WidgetUpdateSchema },
      preHandler: [requireAuth],
    },
    async (request, reply) => {
      const id = Number(request.params.id)
      const existing = getWidget(db, id)
      if (!existing) {
        reply.code(404).send({ error: 'Widget not found' })
        return
      }

      const updateData = { ...request.body }
      if (existing.widget_type === 'weather' && updateData.config) {
        try {
          updateData.config = await resolveWeatherZip(updateData.config)
        } catch (e) {
          if (e instanceof GeocodingError) {
            reply.code(400).send({ error: e.message })
            return
          }
          throw e
        }
      }

      const widget = updateWidget(db, id, updateData)
      return widget
    },
  )

  app.delete<{ Params: { id: string } }>(
    '/api/widgets/:id',
    {
      preHandler: [requireAuth],
    },
    async (request, reply) => {
      const id = Number(request.params.id)
      const existing = getWidget(db, id)
      if (!existing) {
        reply.code(404).send({ error: 'Widget not found' })
        return
      }
      app.log.info(
        `Widget removed: ${existing.widget_type} (id=${existing.id}) from layout ${existing.layout_id}`,
      )
      deleteWidget(db, id)
      reply.code(204).send()
    },
  )

  app.put<{ Params: { layoutId: string }; Body: WidgetPositionUpdate[] }>(
    '/api/layouts/:layoutId/widgets/positions',
    {
      schema: { body: Type.Array(WidgetPositionUpdateSchema) },
      preHandler: [requireAuth],
    },
    async (request, reply) => {
      const layoutId = Number(request.params.layoutId)
      const layout = getLayout(db, layoutId)
      if (!layout) {
        reply.code(404).send({ error: 'Layout not found' })
        return
      }

      // Verify all widget IDs belong to the layout
      const layoutWidgets = getWidgetsByLayout(db, layoutId)
      const layoutWidgetIds = new Set(layoutWidgets.map((w) => w.id))
      for (const pos of request.body) {
        if (!layoutWidgetIds.has(pos.id)) {
          reply
            .code(404)
            .send({ error: `Widget ${pos.id} not found in layout` })
          return
        }
      }

      batchUpdatePositions(db, layoutId, request.body)
      const updated = getWidgetsByLayout(db, layoutId)
      return updated
    },
  )
}
